compiler/src/dotty/tools/dotc/typer/Nullables.scala

```diff
@@ -0,0 +1,341 @@
+        tree match
+          case CompareNull(TrackedRef(ref), testEqual) =>
+            if testEqual then setConditional(Set(), Set(ref))
+            else setConditional(Set(ref), Set())
+          case Apply(Select(x, _), y :: Nil) =>
+            val xc = x.notNullConditional
+            val yc = y.notNullConditional
+            if !(xc.isEmpty && yc.isEmpty) then
+              if tree.symbol == defn.Boolean_&& then
+                setConditional(xc.ifTrue | yc.ifTrue, xc.ifFalse & yc.ifFalse)
+              else if tree.symbol == defn.Boolean_|| then
+                setConditional(xc.ifTrue & yc.ifTrue, xc.ifFalse | yc.ifFalse)
+          case Select(x, _) if tree.symbol == defn.Boolean_! =>
+            val xc = x.notNullConditional
+            if !xc.isEmpty then
+              setConditional(xc.ifFalse, xc.ifTrue)
+          case _ =>
+      tree
+
+    /** Compute nullability information for this tree and all its subtrees */
+    def computeNullableDeeply()(given Context): Unit =
+      new TreeTraverser {
+        def traverse(tree: Tree)(implicit ctx: Context) =
+          traverseChildren(tree)
+          tree.computeNullable()
+      }.traverse(tree)
+
+  given (tree: Assign)
+    def computeAssignNullable()(given Context): tree.type = tree.lhs match
+      case TrackedRef(ref) =>
+        tree.withNotNullInfo(NotNullInfo(Set(), Set(ref))) // TODO: refine with nullability type info
+      case _ => tree
+
+  private val analyzedOps = Set(nme.EQ, nme.NE, nme.eq, nme.ne, nme.ZAND, nme.ZOR, nme.UNARY_!)
+
+  /** A map from (name-) offsets of all local variables in this compilation unit
+   *  that can be tracked for being not null to the list of spans of assignments
+   *  to these variables. A variable can be tracked if it has only reachable assignments.
+   *  An assignment is reachable if the path of tree nodes between the block enclosing
+   *  the variable declaration to the assignment consists only of if-expressions,
+   *  while-expressions, block-expressions and type-ascriptions.
+   *  Only reachable assignments are handled correctly in the nullability analysis.
+   *  Therefore, variables with unreachable assignments can be assumed to be not-null
+   *  only if their type asserts it.
+   */
+  def assignmentSpans(given Context): Map[Int, List[Span]] =
+    import ast.untpd._
+
+    object populate extends UntypedTreeTraverser with
+
+      /** The name offsets of variables that are tracked */
+      var tracked: Map[Int, List[Span]] = Map.empty
+
+      /** Map the names of potentially trackable candidate variables in scope to the spans
+       *  of their reachable assignments
+       */
+      val candidates = mutable.Map[Name, List[Span]]()
+
+      /** An assignment to a variable that's not in reachable makes the variable
+       *  ineligible for tracking
+       */
+      var reachable: Set[Name] = Set.empty
+
+      def traverse(tree: Tree)(implicit ctx: Context) =
```

Comment:
This is nice: we didn't think to drop to untyped trees for detecting whether a variable is captured by a closure.